src/rebalancers/libraries/RebalanceOptimizationMath.sol

```diff
@@ -0,0 +1,296 @@
+     * @param amountIn An approximation of the amount of tokenIn, based on the optimal swap through the pool itself without slippage.
+     * @param amountOut An approximation of the amount of tokenOut, based on the optimal swap through the pool itself without slippage.
+     * @return amountOut The amount of tokenOut.
+     * @dev The optimal amountIn and amountOut are defined such that after the swap the maximum amount of liquidity that can be added to the position.
+     * This means that there are no leftovers of either token0 or token1,
+     * and liquidity0 (calculated via getLiquidityForAmount0) will be exactly equal to liquidity1 (calculated via getLiquidityForAmount1).
+     * @dev The optimal amountIn and amountOut depend on the sqrtPrice of the pool via the liquidity calculations,
+     * but the sqrtPrice in turn depends on the amountIn and amountOut via the swap calculations.
+     * Since both are highly non-linear, this problem is (according to our understanding) not analytically solvable.
+     * Therefore we use an iterative approach to find the optimal swap parameters.
+     * The stop criterium is defined when the relative difference between liquidity0 and liquidity1 is below the convergence threshold.
+     * @dev Convergence is not guaranteed, worst case or the transaction reverts, or a non-optimal swap is performed,
+     * But then minLiquidity enforces that either enough liquidity is minted or the transaction will revert.
+     * @dev We assume constant active liquidity when calculating the swap parameters.
+     * For illiquid pools, or positions that are large relatively to the pool liquidity, this might result in reverting rebalances.
+     * But since a minimum amount of liquidity is enforced, should not lead to loss of principal.
+     */
+    function _getAmountOutWithSlippage(
+        bool zeroToOne,
+        uint256 fee,
+        uint128 usableLiquidity,
+        uint160 sqrtPriceOld,
+        uint160 sqrtRatioLower,
+        uint160 sqrtRatioUpper,
+        uint256 amount0,
+        uint256 amount1,
+        uint256 amountIn,
+        uint256 amountOut
+    ) internal returns (uint256) {
+        emit Log2(sqrtPriceOld, sqrtRatioLower, sqrtRatioUpper);
+        uint160 sqrtPriceNew;
+        bool stopCondition;
+        // We iteratively solve for sqrtPrice, amountOut and amountIn, so that the maximal amount of liquidity can be added to the position.
+        for (uint256 i = 0; i < MAX_ITERATIONS; i++) {
+            emit Log(sqrtPriceNew, amountIn, amountOut);
+            // Find a better approximation for sqrtPrice, given the best approximations for the optimal amountIn and amountOut.
+            sqrtPriceNew = _approximateSqrtPriceNew(zeroToOne, fee, usableLiquidity, sqrtPriceOld, amountIn, amountOut);
+
+            // If the position is out of range, we can calculate the exact solution.
+            if (sqrtPriceNew >= sqrtRatioUpper) {
+                // Position is out of range and fully in token1.
```

Comment:
same as rebalance logic, for clarity: _Rebalance to a single-sided liquidity position in token 1._